File: incident-service/src/main/java/com/emergency/incident/service/IncidentService.java
package com.emergency.incident.service;

import com.emergency.incident.dto.*;
import com.emergency.incident.model.*;
import com.emergency.incident.repository.IncidentRepository;
import com.emergency.incident.repository.ResponderStationRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class IncidentService {

    private static final Logger log = LoggerFactory.getLogger(IncidentService.class);
    private static final double EARTH_RADIUS_KM = 6371.0;

    private final IncidentRepository incidentRepository;
    private final ResponderStationRepository stationRepository;

    public IncidentService(IncidentRepository incidentRepository,
                           ResponderStationRepository stationRepository) {
        this.incidentRepository = incidentRepository;
        this.stationRepository = stationRepository;
    }

    @Transactional
    public IncidentDTO createIncident(CreateIncidentRequest request, Long adminId) {
        Incident incident = Incident.builder()
                .citizenName(request.getCitizenName())
                .citizenPhone(request.getCitizenPhone())
                .incidentType(request.getIncidentType())
                .latitude(request.getLatitude())
                .longitude(request.getLongitude())
                .locationAddress(request.getLocationAddress())
                .notes(request.getNotes())
                .createdBy(adminId)
                .status(IncidentStatus.CREATED)
                .build();

        incident = incidentRepository.save(incident);

        // Auto-assign nearest available responder
        try {
            StationType requiredStationType = determineStationType(request.getIncidentType());
            ResponderStation nearestStation = findNearestAvailableStation(
                    request.getLatitude(), request.getLongitude(), requiredStationType
            );

            if (nearestStation != null) {
                incident.setAssignedUnitId(nearestStation.getId());
                incident.setAssignedUnitName(nearestStation.getName());
                incident.setAssignedUnitType(nearestStation.getStationType().name());
                incident.setStatus(IncidentStatus.DISPATCHED);
                incident.setDispatchedAt(LocalDateTime.now());
                incident = incidentRepository.save(incident);
                log.info("Incident {} auto-assigned to station: {} ({})",
                        incident.getId(), nearestStation.getName(), nearestStation.getStationType());
            } else {
                log.warn("No available station found for incident type: {}", request.getIncidentType());
            }
        } catch (Exception e) {
            log.error("Error auto-assigning responder for incident {}: {}", incident.getId(), e.getMessage());
        }

        return mapToDTO(incident);
    }

    public IncidentDTO getIncidentById(Long id) {
        Incident incident = incidentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Incident not found with id: " + id));
        return mapToDTO(incident);
    }

    public List<IncidentDTO> getOpenIncidents() {
        return incidentRepository.findOpenIncidents().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<IncidentDTO> getAllIncidents() {
        return incidentRepository.findAll().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    @Transactional
    public IncidentDTO updateIncidentStatus(Long id, UpdateStatusRequest request) {
        Incident incident = incidentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Incident not found with id: " + id));

        incident.setStatus(request.getStatus());

        if (request.getStatus() == IncidentStatus.RESOLVED) {
            incident.setResolvedAt(LocalDateTime.now());
        } else if (request.getStatus() == IncidentStatus.DISPATCHED) {
            incident.setDispatchedAt(LocalDateTime.now());
        }

        incident = incidentRepository.save(incident);
        return mapToDTO(incident);
    }

    @Transactional
    public IncidentDTO assignUnit(Long incidentId, AssignUnitRequest request) {
        Incident incident = incidentRepository.findById(incidentId)
                .orElseThrow(() -> new RuntimeException("Incident not found with id: " + incidentId));

        ResponderStation station = stationRepository.findById(request.getUnitId())
                .orElseThrow(() -> new RuntimeException("Station not found with id: " + request.getUnitId()));

        incident.setAssignedUnitId(station.getId());
        incident.setAssignedUnitName(station.getName());
        incident.setAssignedUnitType(station.getStationType().name());
        incident.setStatus(IncidentStatus.DISPATCHED);
        incident.setDispatchedAt(LocalDateTime.now());

        incident = incidentRepository.save(incident);
        return mapToDTO(incident);
    }

    /**
     * Determines the required station type based on incident type.
     */
    private StationType determineStationType(IncidentType incidentType) {
        return switch (incidentType) {
            case ROBBERY, CRIME, ASSAULT -> StationType.POLICE_STATION;
            case FIRE -> StationType.FIRE_STATION;
            case MEDICAL_EMERGENCY, ACCIDENT -> StationType.HOSPITAL;
            case NATURAL_DISASTER -> StationType.FIRE_STATION;
            case OTHER -> StationType.POLICE_STATION;
        };
    }

    /**
     * Finds the nearest available station of the required type using the Haversine formula.
     */
    private ResponderStation findNearestAvailableStation(double incidentLat, double incidentLon, StationType stationType) {
        List<ResponderStation> availableStations = stationRepository.findByStationTypeAndIsAvailableTrue(stationType);

        if (availableStations.isEmpty()) {
            return null;
        }

        return availableStations.stream()
                .min(Comparator.comparingDouble(station ->
                        calculateHaversineDistance(incidentLat, incidentLon, station.getLatitude(), station.getLongitude())
                ))
                .orElse(null);
    }

    /**
     * Calculates the distance between two geographic coordinates using the Haversine formula.
     * Returns distance in kilometers.
     */
    private double calculateHaversineDistance(double lat1, double lon1, double lat2, double lon2) {
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);

        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
                + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
                * Math.sin(dLon / 2) * Math.sin(dLon / 2);

        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

        return EARTH_RADIUS_KM * c;
    }

    private IncidentDTO mapToDTO(Incident incident) {
        return IncidentDTO.builder()
                .id(incident.getId())
                .citizenName(incident.getCitizenName())
                .citizenPhone(incident.getCitizenPhone())
                .incidentType(incident.getIncidentType())
                .latitude(incident.getLatitude())
                .longitude(incident.getLongitude())
                .locationAddress(incident.getLocationAddress())
                .notes(incident.getNotes())
                .createdBy(incident.getCreatedBy())
                .assignedUnitId(incident.getAssignedUnitId())
                .assignedUnitName(incident.getAssignedUnitName())
                .assignedUnitType(incident.getAssignedUnitType())
                .status(incident.getStatus())
                .timestamp(incident.getTimestamp())
                .dispatchedAt(incident.getDispatchedAt())
                .resolvedAt(incident.getResolvedAt())
                .build();
    }
}

File: nodejs-migration/incident-service/src/services/IncidentService.ts
import { AppDataSource } from '../database/database';
import { Incident, IncidentStatus, IncidentType } from '../entities/Incident';
import { ResponderStation, StationType } from '../entities/ResponderStation';
import {
  CreateIncidentRequest,
  UpdateStatusRequest,
  AssignUnitRequest,
  IncidentDTO,
} from '../dto/incident.dto';

export class IncidentService {
  private incidentRepository = AppDataSource.getRepository(Incident);
  private stationRepository = AppDataSource.getRepository(ResponderStation);

  private readonly EARTH_RADIUS_KM = 6371.0;

  async createIncident(
    request: CreateIncidentRequest,
    adminId: number
  ): Promise<IncidentDTO> {
    const incident = this.incidentRepository.create({
      citizenName: request.citizenName,
      citizenPhone: request.citizenPhone,
      incidentType: request.incidentType,
      latitude: request.latitude,
      longitude: request.longitude,
      locationAddress: request.locationAddress,
      notes: request.notes,
      createdBy: adminId,
      status: IncidentStatus.CREATED,
    });

    await this.incidentRepository.save(incident);

    // Auto-assign nearest available responder
    try {
      const requiredStationType = this.determineStationType(
        request.incidentType
      );
      const nearestStation = await this.findNearestAvailableStation(
        request.latitude,
        request.longitude,
        requiredStationType
      );

      if (nearestStation) {
        await this.assignStationToIncident(incident, nearestStation);
        console.log(
          `Incident ${incident.id} auto-assigned to station: ${nearestStation.name}`
        );
      } else {
        console.warn(
          `No available station found for incident type: ${request.incidentType}`
        );
      }
    } catch (error) {
      console.error(`Error auto-assigning responder for incident: ${error}`);
    }

    return this.mapToDTO(incident);
  }

  private determineStationType(incidentType: IncidentType): StationType {
    switch (incidentType) {
      case IncidentType.FIRE:
      case IncidentType.HAZMAT:
        return StationType.FIRE;
      case IncidentType.MEDICAL:
        return StationType.HOSPITAL;
      case IncidentType.ACCIDENT:
        return StationType.POLICE;
      default:
        return StationType.POLICE;
    }
  }

  private calculateDistance(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number
  ): number {
    const dLat = this.degreesToRadians(lat2 - lat1);
    const dLon = this.degreesToRadians(lon2 - lon1);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this.degreesToRadians(lat1)) *
        Math.cos(this.degreesToRadians(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);

    const c = 2 * Math.asin(Math.sqrt(a));
    return this.EARTH_RADIUS_KM * c;
  }

  private degreesToRadians(degrees: number): number {
    return (degrees * Math.PI) / 180;
  }

  private async findNearestAvailableStation(
    latitude: number,
    longitude: number,
    stationType: StationType
  ): Promise<ResponderStation | null> {
    const stations = await this.stationRepository.find({
      where: { stationType, isAvailable: true },
    });

    if (stations.length === 0) return null;

    let nearestStation = stations[0];
    let minDistance = this.calculateDistance(
      latitude,
      longitude,
      parseFloat(stations[0].latitude.toString()),
      parseFloat(stations[0].longitude.toString())
    );

    for (const station of stations) {
      const distance = this.calculateDistance(
        latitude,
        longitude,
        parseFloat(station.latitude.toString()),
        parseFloat(station.longitude.toString())
      );

      if (distance < minDistance) {
        minDistance = distance;
        nearestStation = station;
      }
    }

    return nearestStation;
  }

  private async assignStationToIncident(
    incident: Incident,
    station: ResponderStation
  ): Promise<void> {
    incident.assignedUnitId = station.id;
    incident.assignedUnitName = station.name;
    incident.assignedUnitType = station.stationType;
    incident.status = IncidentStatus.DISPATCHED;
    incident.dispatchedAt = new Date();

    await this.incidentRepository.save(incident);

    // Update hospital occupancy if applicable
    if (station.stationType === StationType.HOSPITAL) {
      station.currentOccupancy =
        (station.currentOccupancy || 0) + 1;
      await this.stationRepository.save(station);
    }
  }

  async getIncidentById(id: number): Promise<IncidentDTO> {
    const incident = await this.incidentRepository.findOne({
      where: { id },
    });

    if (!incident) {
      throw new Error(`Incident not found with id: ${id}`);
    }

    return this.mapToDTO(incident);
  }

  async getOpenIncidents(): Promise<IncidentDTO[]> {
    const incidents = await this.incidentRepository.find({
      where: [
        { status: IncidentStatus.CREATED },
        { status: IncidentStatus.DISPATCHED },
        { status: IncidentStatus.IN_PROGRESS },
      ],
    });

    return incidents.map((i) => this.mapToDTO(i));
  }

  async getAllIncidents(): Promise<IncidentDTO[]> {
    const incidents = await this.incidentRepository.find();
    return incidents.map((i) => this.mapToDTO(i));
  }

  async updateIncidentStatus(
    id: number,
    request: UpdateStatusRequest
  ): Promise<IncidentDTO> {
    const incident = await this.incidentRepository.findOne({
      where: { id },
    });

    if (!incident) {
      throw new Error(`Incident not found with id: ${id}`);
    }

    incident.status = request.status;
    await this.incidentRepository.save(incident);

    return this.mapToDTO(incident);
  }

  async assignUnit(
    id: number,
    request: AssignUnitRequest
  ): Promise<IncidentDTO> {
    const incident = await this.incidentRepository.findOne({
      where: { id },
    });

    if (!incident) {
      throw new Error(`Incident not found with id: ${id}`);
    }

    incident.assignedUnitId = request.unitId;
    incident.assignedUnitName = request.unitName;
    incident.assignedUnitType = request.unitType;
    incident.status = IncidentStatus.DISPATCHED;
    incident.dispatchedAt = new Date();

    await this.incidentRepository.save(incident);

    return this.mapToDTO(incident);
  }

  private mapToDTO(incident: Incident): IncidentDTO {
    return {
      id: incident.id,
      citizenName: incident.citizenName,
      citizenPhone: incident.citizenPhone,
      incidentType: incident.incidentType,
      latitude: incident.latitude,
      longitude: incident.longitude,
      locationAddress: incident.locationAddress,
      notes: incident.notes,
      status: incident.status,
      assignedUnitId: incident.assignedUnitId,
      assignedUnitName: incident.assignedUnitName,
      assignedUnitType: incident.assignedUnitType,
      createdAt: incident.createdAt,
      dispatchedAt: incident.dispatchedAt,
      updatedAt: incident.updatedAt,
    };
  }
}
